Additional preprocessors

matejklemen opened the pull request:
<!-- This is a pull request template. Please fill in the relevant details in the sections below. -->
##### Description of Changes
<!-- Don't forget to add reference to an existing issue if present. -->
Addresses #7. Implements 3 new preprocessors:
1. Binarizer: convert **numerical columns** into binary columns based on a threshold. Works with either a `Double` scalar or a `DenseVector[Double]`. In the second case, each value of vector represents a threshold for a separate column, e.g. 1st value is used as threshold for 1st numeric column.  
Motivated by [sklearn's Binarizer](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.Binarizer.html).

2. Normalizer: converts **rows** to unit norm. Currently, there are 3 norms supported (L1, L2, max). These should probably be exposed in some kind of distance module for further code reuse, I am open to ideas on where to put these norm functions.  
Motivated by [sklearn's Normalizer](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.Normalizer.html).

3. RangeScaler: converts **numerical columns** to a certain range, e.g. from 0 to 1.  
Motivated by [sklearn's MinMaxScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html) (_RangeScaler_ or _IntervalScaler_ seems like a more intuitive name to me, though I can change it if min-max scaling is a "standardized" name for this).



##### Includes
<!-- Mark the changes included in the PR. -->
- [X] Code changes
- [X] Tests
- [ ] Documentation

inejc commented:
Thanks @matejklemen 👍, will look at the PR as soon as I find time for it.

matejklemen commented:
Alright, no rush.

I've fixed the compilation error that was previously present and tried to address Codacy's issues with the PR, though it seems that I don't understand what the issue _"Found unnamed result or parameter"_ actually means :sweat_smile:

inejc commented on src/main/scala/io/picnicml/doddlemodel/preprocessing/Normalizer.scala:
```diff
@@ -0,0 +1,35 @@
+  }
+
+  implicit lazy val ev: Transformer[Normalizer] = new Transformer[Normalizer] {
+    override def fit(model: Normalizer, x: Features): Normalizer = model
```
Newline is used between `new X {` and the next line throughout the project.

inejc commented on src/main/scala/io/picnicml/doddlemodel/preprocessing/Normalizer.scala:
```diff
@@ -0,0 +1,35 @@
+      case "max" => (x: Features) => max(abs(x), Axis._1)
+      case _ => throw new IllegalArgumentException("Unsupported norm")
+    }
+    new Normalizer(normFunction)
```
Same as above, `new` is not needed.

inejc commented on src/main/scala/io/picnicml/doddlemodel/preprocessing/RangeScaler.scala:
```diff
@@ -0,0 +1,52 @@
+package io.picnicml.doddlemodel.preprocessing
+
+import breeze.linalg.{*, Axis, max, min}
```
Please use `Code -> Optimize Imports` if using IntelliJ (not sure what the command is in other IDEs) to reorder imports lexicographically.

inejc commented on src/main/scala/io/picnicml/doddlemodel/preprocessing/RangeScaler.scala:
```diff
@@ -0,0 +1,52 @@
+    override def isFitted(model: RangeScaler): Boolean =
+      model.scale.isDefined && model.minAdjustment.isDefined
+  }
+
```
This newline should be removed.

inejc commented on src/main/scala/io/picnicml/doddlemodel/preprocessing/RangeScaler.scala:
```diff
@@ -0,0 +1,52 @@
+  def apply(range: (Double, Double), featureIndex: FeatureIndex): RangeScaler = {
+    val (lowerBound, upperBound) = range
+    val numNumeric = featureIndex.numerical.columnIndices.length
+    require(numNumeric > 0, "There must be at least 1 numeric column in the given data")
```
This is handled differently in [the standard scaler case](https://github.com/picnicml/doddle-model/blob/master/src/main/scala/io/picnicml/doddlemodel/preprocessing/StandardScaler.scala#L42); we don't throw the exception if there are no numeric columns and simply no columns are preprocessed (although we should display a warning in that case which is missing there). Not sure what the better way is.

matejklemen replied:
Okay, I'll change this - I think both ways could work, so I guess we just make the behavior consistent across preprocessors for now.

Just making sure, this comment applies to `Binarizer` as well, right? (currently, it throws an exception if there are no numerical features in data)

inejc replied:
I agree, let's go for consistency for now; we can change this later in all files. Yup, it applies to `Binarizer` as well.

inejc commented on src/main/scala/io/picnicml/doddlemodel/preprocessing/RangeScaler.scala:
```diff
@@ -0,0 +1,52 @@
+      model.copy(scale.some, minAdjustment.some)
+    }
+
+    override protected def transformSafe(model: RangeScaler, x: Features): Features = {
```
The same as above: preprocessing shouldn't drop any columns from the data.

inejc commented on src/test/scala/io/picnicml/doddlemodel/preprocessing/RangeScalerTest.scala:
```diff
@@ -0,0 +1,31 @@
+      List(0.0, 1.0)
+    )) shouldBe true
+  }
+
```
Newline should be removed.

inejc commented on src/test/scala/io/picnicml/doddlemodel/preprocessing/RangeScalerTest.scala:
```diff
@@ -0,0 +1,31 @@
+class RangeScalerTest extends FlatSpec with Matchers with TestingUtils {
+  implicit val doubleTolerance: Equality[Double] = TolerantNumerics.tolerantDoubleEquality(1e-4)
+
+  "Range scaler" should "scale features to specified range" in {
```
Once the preprocessor is implemented in a way so that all columns are retained in the output, we would need two different tests; one that transforms all numerical columns (`RangeScaler((0.0, 1.0), featureIndex)`) and one with a subset of numerical columns (`RangeScaler((0.0, 1.0), featureIndex)`). Take a look at the standard scaler tests for examples.

matejklemen commented on src/main/scala/io/picnicml/doddlemodel/preprocessing/Binarizer.scala:
```diff
@@ -0,0 +1,38 @@
+import io.picnicml.doddlemodel.data.{Features, RealVector}
+import io.picnicml.doddlemodel.typeclasses.Transformer
+
+case class Binarizer private (private val featureIndex: FeatureIndex,
```
It's a good point, I actually put it like that originally, but there was a compilation error because the case class auto-generates an `apply` method with same signature as the one I create inside `Binarizer` companion object. Here's the copy-paste of the error from CI:
```
[info] Non-compiled module 'compiler-bridge_2.11' for Scala 2.11.12. Compiling...
[info]   Compilation completed in 7.738s.
[error] /home/circleci/repo/src/main/scala/io/picnicml/doddlemodel/preprocessing/Binarizer.scala:8:12: method apply is defined twice
[error]   conflicting symbols both originated in file '/home/circleci/repo/src/main/scala/io/picnicml/doddlemodel/preprocessing/Binarizer.scala'
[error] case class Binarizer private (private val thresholds: RealVector,
[error]            ^
[error] one error found
[error] (Compile / compileIncremental) Compilation failed
```

Do you have an idea on how to fix this problem in a nicer way?

inejc replied:
Hmm weird, this compiles for me:
```scala
case class Binarizer private (private val thresholds: RealVector, private val featureIndex: FeatureIndex)

object Binarizer {

  def apply(threshold: Double, featureIndex: FeatureIndex): Binarizer = {
    val numNumeric: Int = featureIndex.numerical.columnIndices.length
    require(numNumeric > 0, "There must be at least 1 numeric column in the given data")
    val thresholdsExtended = DenseVector.fill(numNumeric) {threshold}
    Binarizer(thresholdsExtended, featureIndex)
  }

  def apply(thresholds: RealVector, featureIndex: FeatureIndex): Binarizer = {
    val numNumeric = featureIndex.numerical.columnIndices.length
    require(numNumeric > 0, "There must be at least 1 numeric column in the given data")
    require(numNumeric == thresholds.length, "A threshold should be given for every numerical column")
    Binarizer(thresholds, featureIndex)
  }
  ...
}
```
Btw no need to put definitions in multiple lines if they fit in 120 characters, take a look at [this](https://github.com/picnicml/doddle-model#development).

matejklemen replied:
You are right about that, it does compile. However, making the above change introduces a problem - some Binarizer tests do not pass (or end, for that matter). The problem seems to be that the `Binarizer(thresholds, featureIndex)` call in the second `apply` function calls itself instead of the `apply` method that is automatically created by the case class (as they have same signatures).

Introducing the `new` keyword (redundant, but I thought it could fix things) fixes the problem, but does not compile on Scala 2.11.12:
```
[info] Compiling 48 Scala sources to /home/matej/Documents/doddle-model/target/scala-2.11/classes ...
[error] /home/matej/Documents/doddle-model/src/main/scala/io/picnicml/doddlemodel/preprocessing/Binarizer.scala:16:5: ambiguous reference to overloaded definition,
[error] both method apply in object Binarizer of type (thresholds: io.picnicml.doddlemodel.data.RealVector, featureIndex: io.picnicml.doddlemodel.data.Feature.FeatureIndex)io.picnicml.doddlemodel.preprocessing.Binarizer
[error] and  method apply in object Binarizer of type (thresholds: io.picnicml.doddlemodel.data.RealVector, featureIndex: io.picnicml.doddlemodel.data.Feature.FeatureIndex)io.picnicml.doddlemodel.preprocessing.Binarizer
[error] match argument types (breeze.linalg.DenseVector[Double],io.picnicml.doddlemodel.data.Feature.FeatureIndex) and expected result type io.picnicml.doddlemodel.preprocessing.Binarizer
[error]     Binarizer(thresholdsExtended, featureIndex)
[error]     ^
[error] /home/matej/Documents/doddle-model/src/main/scala/io/picnicml/doddlemodel/preprocessing/Binarizer.scala:8:12: method apply is defined twice
[error]   conflicting symbols both originated in file '/home/matej/Documents/doddle-model/src/main/scala/io/picnicml/doddlemodel/preprocessing/Binarizer.scala'
[error] case class Binarizer private (private val thresholds: RealVector, private val featureIndex: FeatureIndex)
[error]            ^
[error] two errors found
```

inejc replied:
Oh I see what the problem is. In that case, it makes sense to simply make the constructor public and remove the "overloaded" apply method:
```scala
case class Binarizer(private val thresholds: RealVector, private val featureIndex: FeatureIndex)

object Binarizer {
  // only keep this apply method
  def apply(threshold: Double, featureIndex: FeatureIndex): Binarizer = {
    val numNumeric: Int = featureIndex.numerical.columnIndices.length
    require(numNumeric > 0, "There must be at least 1 numeric column in the given data")
    val thresholdsExtended = DenseVector.fill(numNumeric) {threshold}
    Binarizer(thresholdsExtended, featureIndex)
  }
 ...
}
```
If after refactoring you still need to validate the input arguments you can do that directly in the case class definition:
```scala
case class Binarizer(private val thresholds: RealVector, private val featureIndex: FeatureIndex) {
  require(something > 0, "'something' has to be positive.")
}
```

matejklemen commented:
Alright, I think I got all of them. Thanks for the helpful comments :slightly_smiling_face: 

I was a bit confused about sealed traits, so I did it slightly differently than what you said. It seemed redundant to me to define available `Norm`s twice - once in `Normalizer` and again in `Norms` so I just did it in `Norms`.
Please let me know if that's ok.

inejc commented on src/main/scala/io/picnicml/doddlemodel/preprocessing/RangeScaler.scala:
```diff
@@ -0,0 +1,58 @@
+      val numericColIndices = model.featureIndex.numerical.columnIndices
+      val numericColsOnly = x(::, numericColIndices).toDenseMatrix
+      val (colMax: RealVector, colMin: RealVector) =
+        (max(numericColsOnly, Axis._0).inner, min(numericColsOnly, Axis._0).inner)
```
I've never seen `.inner` before, is it for transposing? `.t` can also be used in that case.

inejc commented on src/main/scala/io/picnicml/doddlemodel/preprocessing/RangeScaler.scala:
```diff
@@ -0,0 +1,58 @@
+    override def fit(model: RangeScaler, x: Features): RangeScaler = {
+      val (lowerBound, upperBound) = model.range
+      val numericColIndices = model.featureIndex.numerical.columnIndices
+      val numericColsOnly = x(::, numericColIndices).toDenseMatrix
```
I think we can simply remove `.toDenseMatrix` here and then write `(max(x(::, numericColIndices), Axis._0).inner, min(x(::, numericColIndices), Axis._0).inner)`.

matejklemen replied:
Good catch. One small problem remains though - the line `val dataRange = colMax - colMin` somehow creates a `Vector[Double]` out of two `DenseVector[Double]`s, making the following line fail (because `:==` is not possible between breeze's `Vector` and `Double`):
`dataRange(dataRange :== 0.0) := 1.0`.

I see two ways to fix this, please let me know which one sounds better/ if there's a better way (seems like there should be):
1.) Type cast, i.e. `val dataRange = (colMax - colMin).asInstanceOf[DenseVector[Double]]`  
- does not look very nice, pretty sure explicit type casts are to be avoided if possible
- not sure how this is with copying (how much copying does it do, I don't see how it could be in-place)

2.) `toDenseVector`  
- same problems as `toDenseMatrix`, but vectors are 1D and therefore have a smaller memory footprint, so not as much data would be copied.

inejc replied:
Hmm `val dataRange = colMax - colMin` actually results in `dataRange: DenseVector[Double]` as far as I can see (tests are also green). Try running the code and see if you actually observe `dataRange: Vector[Double]` (maybe it's reported by your IDE wrongly).

matejklemen replied:
It seems you were right, it works. I was not aware I could be tricked by the REPL as well as the IDE :sweat_smile:

inejc replied:
Weird indeed. Which IDE are you using?

matejklemen replied:
A quite stale version of IntelliJ Ultimate edition (2018.2.4). It seems like it's time to finally update it :sweat_smile:

inejc replied:
Ok, this is the only thing that still doesn't compile for me 😅. This compiles, however:
```scala
val colMax = max(x(::, numericColIndices), Axis._0).t.toDenseVector
val colMin = min(x(::, numericColIndices), Axis._0).t.toDenseVector
val dataRange = colMax - colMin
```
it's still better than the previous `.toDenseMatrix` as it only materializes two vectors of length `x(::, numericColIndices).cols`. I belive the max and min actually returns a `Vector[Double]` because we are passing in a `SliceMatrix`. Anyway, the above should work as intended.

matejklemen replied:
Fixed. Weird that the tests passed anyway, are you using a version of breeze that is different from the one in project/CI? Anyways, good that you caught the problem before merging :smiley:

inejc replied:
That shouldn't be the case as I checked out your branch. I'm not sure why it didn't fail, I will have to investigate 🙂.

inejc commented on src/main/scala/io/picnicml/doddlemodel/preprocessing/RangeScaler.scala:
```diff
@@ -0,0 +1,58 @@
+      val numericColIndices = model.featureIndex.numerical.columnIndices
+      // only perform scaling if there are numerical columns, otherwise keep input
+      if(numericColIndices.nonEmpty) {
+        val numericColsOnly = x(::, numericColIndices).toDenseMatrix
```
Is it possible to also rewrite this so that copying is avoided? `.toDenseMatrix` should really be used carefully 🙂.

inejc commented on src/test/scala/io/picnicml/doddlemodel/preprocessing/RangeScalerTest.scala:
```diff
@@ -0,0 +1,56 @@
+import org.scalactic.{Equality, TolerantNumerics}
+import org.scalatest.{FlatSpec, Matchers}
+
+class RangeScalerTest extends FlatSpec with Matchers with TestingUtils {
```
Perhaps another test with a different range would be useful.

matejklemen commented on src/main/scala/io/picnicml/doddlemodel/preprocessing/Binarizer.scala:
```diff
@@ -0,0 +1,40 @@
+
+    override def fit(model: Binarizer, x: Features): Binarizer = model
+
+    override protected def transformSafe(model: Binarizer, x: Features): Features = {
```
Ah, I should have known there was a good reason the code in other preprocessors looked more complex :sweat_smile: 

You are right, I will fix it like you suggested. One question though, is the `zipWithIndex` really necessary? As far as I can tell, `colIndex` and `thresholdIndex` will always be the same.

inejc replied:
Columns referenced in `featureIndex` and columns in the actual data matrix might not be the same. You can think of `featureIndex` as a way of saying "we are talking about these very specific columns" (e.g. data matrix `x` has 4 columns and `featureIndex.columnIndices` is `IndexedSeq(0, 2)` meaning we are only talking about the first and the third column). Namely, an instance of `featureIndex` is passed to preprocessors and enables the ability to preprocess only a subset of columns. This should be implemented in a more explicit manner in the future, I know :).

As a result `colIndex` and `thresholdIndex` shouldn't alway be the same above. Even if an instance of `featureIndex` that specifies all columns is passed to the preprocessor, the categorical columns are ignored by `.numerical` which will result in `colIndex` and `thresholdIndex` being out of sync, because `thresholds` only contains thresholds for numerical columns. Hope that makes sense, LMK if I'm being unclear.

matejklemen commented:
New batch of changes:
- added tests for Norms (my bad, completely forgot before :sweat_smile: )
- made minor style fixes
- added docs for the three preprocessors (I only put the imports that I think are non-obvious in the examples, as I'm doing in #69)
- added a test with different range for `RangeScaler`
- removed unnecessary copying that is done by `toDenseMatrix`

inejc commented:
@matejklemen everything LGTM except the comment I wrote regarding compilation (a single and simple change). I will merge once this one is pushed 😄.
